### PR TITLE
alcs 75 - Application Decision Meeting Part 1

### PR DESCRIPTION
**Pull request by mhuseinov**
* Application Decision Meeting
    * entity
    * service
    * controller 
* Review page with Decision Meetings
* unit tests

### CODE REVIEW

File: alcs-frontend/src/app/features/application/application.component.scss [hunk 1]
```diff
@@ -6,8 +6,10 @@
 
 .application {
   margin: 12px;
+  width:100%;
```
Comment (dhaselhan) [inline alcs-frontend/src/app/features/application/application.component.scss:9]: spacing is off

File: alcs-frontend/src/app/features/application/application.component.scss [hunk 2]
```diff
@@ -6,8 +6,10 @@
...
 }
 
 .child-content {
   margin: 12px 0;
+
```
Comment (dhaselhan) [inline alcs-frontend/src/app/features/application/application.component.scss:14]: same here

File: alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import { BehaviorSubject, firstValueFrom } from 'rxjs';
+import { environment } from '../../../../environments/environment';
+import { ToastService } from '../../toast/toast.service';
+import { ApplicationDecisionMeetingDto, CreateApplicationDecisionMeetingDto } from './application-decision-meeting.dto';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class ApplicationDecisionMeetingService {
+  public $decisionMeetings = new BehaviorSubject<ApplicationDecisionMeetingDto[]>([]);
```
Comment (dhaselhan) [inline alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts:13]: no need to use public modifier its impicit

File: alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  private url = `${environment.apiRoot}/application-decision-meeting`;
+
+  constructor(private http: HttpClient, private toastService: ToastService) {}
+
+  async fetch(fileNumber: string) {
+    let meetings = [] as ApplicationDecisionMeetingDto[];
```
Comment (dhaselhan) [inline alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts:19]: let meetings: ApplicationDecisionMeetingDto[] = []; is more preferred as it's specifying the type rather than casting.

File: alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        })
+      );
+      await this.fetch(decisionMeeting.applicationFileNumber);
+    } catch (e) {
+      this.toastService.showErrorToast('Failed to update decision meeting');
+    }
+  }
+
+  async create(decisionMeeting: CreateApplicationDecisionMeetingDto) {
+    let meeting;
```
Comment (dhaselhan) [inline alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts:45]: remove this

File: alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+          ...decisionMeeting,
+          date: dayjs(decisionMeeting.date).startOf('day').add(12, 'hours').valueOf(),
+        })
+      );
+      await this.fetch(decisionMeeting.applicationFileNumber);
+    } catch (e) {
+      this.toastService.showErrorToast('Failed to create decision meeting');
+    }
+
+    return meeting;
```
Comment (dhaselhan) [inline alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts:58]: remove this

File: alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  async create(decisionMeeting: CreateApplicationDecisionMeetingDto) {
+    let meeting;
+    try {
+      meeting = await firstValueFrom(
+        this.http.post<ApplicationDecisionMeetingDto>(this.url, {
+          ...decisionMeeting,
+          date: dayjs(decisionMeeting.date).startOf('day').add(12, 'hours').valueOf(),
+        })
+      );
+      await this.fetch(decisionMeeting.applicationFileNumber);
```
Comment (dhaselhan) [inline alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts:53]: return await this.fetch()

File: alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    } catch (e) {
+      this.toastService.showErrorToast('Failed to create decision meeting');
+    }
+
+    return meeting;
+  }
+
+  async fetchOne(uuid: string) {
+    let meeting = {} as ApplicationDecisionMeetingDto;
```
Comment (dhaselhan) [inline alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts:62]: remove this

File: alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+    try {
+      meeting = await firstValueFrom(this.http.get<ApplicationDecisionMeetingDto>(`${this.url}/meeting/${uuid}`));
+    } catch (err) {
+      this.toastService.showErrorToast('Failed to fetch decision meetings');
+    }
+
+    return meeting;
```
Comment (dhaselhan) [inline alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts:70]: remove this

File: alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  async fetchOne(uuid: string) {
+    let meeting = {} as ApplicationDecisionMeetingDto;
+
+    try {
+      meeting = await firstValueFrom(this.http.get<ApplicationDecisionMeetingDto>(`${this.url}/meeting/${uuid}`));
```
Comment (dhaselhan) [inline alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts:65]: return here

File: alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    } catch (err) {
+      this.toastService.showErrorToast('Failed to fetch decision meetings');
+    }
+
+    return meeting;
+  }
+
+  async delete(uuid: string) {
+    let meeting = {} as ApplicationDecisionMeetingDto;
```
Comment (dhaselhan) [inline alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts:74]: same comments as above

File: alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+    try {
+      meeting = await firstValueFrom(this.http.delete<ApplicationDecisionMeetingDto>(`${this.url}/${uuid}`));
+      this.toastService.showSuccessToast('Meeting deleted');
+    } catch (err) {
+      this.toastService.showErrorToast('Failed to delete meeting');
+    }
+
+    return meeting;
```
Comment (dhaselhan) [inline alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts:83]: I would prefer undefined returned if something goes wrong instead of empty object that has the right type.

File: alcs-frontend/src/app/features/application/review/review.component.ts [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+  selector: 'app-review',
+  templateUrl: './review.component.html',
+  styleUrls: ['./review.component.scss'],
+})
+export class ReviewComponent implements OnInit {
```
Comment (dhaselhan) [inline alcs-frontend/src/app/features/application/review/review.component.ts:8]: You can remove the implement OnInit if its not doing anything and delete the function

File: alcs-frontend/src/app/features/application/review/decision-meeting/decision-meeting.component.ts [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+export class DecisionMeetingComponent implements OnInit {
+  displayedColumns: string[] = ['date', 'action'];
+  decisionMeetings: ApplicationDecisionMeetingDto[] = [];
+
+  @Input()
+  fileNumber: string = '';
+
+  constructor(
+    public dialog: MatDialog,
+    public decisionMeetingService: ApplicationDecisionMeetingService,
```
Comment (dhaselhan) [inline alcs-frontend/src/app/features/application/review/decision-meeting/decision-meeting.component.ts:22]: Should this be private?

File: alcs-frontend/src/app/features/application/review/decision-meeting/decision-meeting.component.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+      maxWidth: '800px',
+      width: '70%',
+      data: {
+        fileNumber: this.fileNumber,
+      },
+    });
+  }
+
+  async onEdit(uuid: string) {
+    const meeting = await this.decisionMeetingService.fetchOne(uuid);
```
Comment (dhaselhan) [inline alcs-frontend/src/app/features/application/review/decision-meeting/decision-meeting.component.ts:45]: Is this refetch to be careful of others editing the meeting?

File: alcs-frontend/src/app/features/application/review/decision-meeting/decision-meeting.component.html [hunk 16]
```diff
@@ -0,0 +1,37 @@
+<br />
```
Comment (dhaselhan) [inline alcs-frontend/src/app/features/application/review/decision-meeting/decision-meeting.component.html:1]: Why is this here? If this is for spacing you should add margin-top to the div below instead.

File: alcs-frontend/src/app/features/application/review/decision-meeting-dialog/decision-meeting-dialog.component.ts [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+      this.model = {
+        ...data,
+        date: new Date(data.date),
+      };
+    } else {
+      this.model = new ApplicationDecisionMeetingForm(data.fileNumber, new Date());
+    }
+  }
+
+  ngOnInit(): void {}
```
Comment (dhaselhan) [inline alcs-frontend/src/app/features/application/review/decision-meeting-dialog/decision-meeting-dialog.component.ts:34]: Remove empty onInit

File: alcs-frontend/src/app/features/application/review/decision-meeting-dialog/decision-meeting-dialog.component.ts [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+  templateUrl: './decision-meeting-dialog.component.html',
+  styleUrls: ['./decision-meeting-dialog.component.scss'],
+})
+export class DecisionMeetingDialogComponent implements OnInit {
+  model: ApplicationDecisionMeetingForm;
+
+  constructor(
+    @Inject(MAT_DIALOG_DATA) public data: ApplicationDecisionMeetingForm,
+    private dialogRef: MatDialogRef<DecisionMeetingDialogComponent>,
+    public decisionMeetingService: ApplicationDecisionMeetingService,
```
Comment (dhaselhan) [inline alcs-frontend/src/app/features/application/review/decision-meeting-dialog/decision-meeting-dialog.component.ts:21]: Should be private?

File: alcs-frontend/src/app/features/application/review/decision-meeting-dialog/decision-meeting-dialog.component.html [hunk 19]
```diff
@@ -0,0 +1,20 @@
+<div mat-dialog-title>
+  <h2 class="card-title">Create Decision Meeting</h2>
```
Comment (dhaselhan) [inline alcs-frontend/src/app/features/application/review/decision-meeting-dialog/decision-meeting-dialog.component.html]: Suggestion:  Change this to Edit Decision Meeting when you are editing.

File: alcs-frontend/src/app/features/application/application.component.html [hunk 20]
```diff
@@ -5,7 +5,7 @@
   <div class="application">
     <h2>{{ application.fileNumber }} ({{ application.applicant }})</h2>
     <div class="child-content">
-      <router-outlet></router-outlet>
+      <router-outlet (activate)="onOutletLoaded($event)"></router-outlet>
```
Comment (dhaselhan) [inline alcs-frontend/src/app/features/application/application.component.html:8]: 👍

File: alcs-api/src/application/application-decision-meeting/application-decision-meeting.service.ts [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+
+@Injectable()
+export class ApplicationDecisionMeetingService {
+  constructor(
+    @InjectRepository(ApplicationDecisionMeeting)
+    private appDecisionMeetingRepository,
+    private applicationService: ApplicationService,
+  ) {}
+
+  async getByAppFileNumber(number) {
```
Comment (dhaselhan) [inline alcs-api/src/application/application-decision-meeting/application-decision-meeting.service.ts:15]: add type to function number: string

File: alcs-api/src/application/application-decision-meeting/application-decision-meeting.service.ts [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+        );
+      }
+    }
+
+    const updatedMeeting = Object.assign(
+      existingMeeting || new ApplicationDecisionMeeting(),
+      decisionMeeting,
+    );
+
+    return await this.appDecisionMeetingRepository.save(updatedMeeting);
```
Comment (dhaselhan) [inline alcs-api/src/application/application-decision-meeting/application-decision-meeting.service.ts:51]: You can remove the await

File: alcs-api/src/application/application-decision-meeting/application-decision-meeting.service.spec.ts [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+import { ApplicationService } from '../application.service';
+import { ApplicationDecisionMeeting } from './application-decision-meeting.entity';
+import { ApplicationDecisionMeetingService } from './application-decision-meeting.service';
+
+describe('ApplicationDecisionMeetingService', () => {
+  let service: ApplicationDecisionMeetingService;
+  let mockAppDecisionMeetingRepository: MockType<
+    Repository<ApplicationDecisionMeeting>
+  >;
+  let mockApplicationService;
```
Comment (dhaselhan) [inline alcs-api/src/application/application-decision-meeting/application-decision-meeting.service.spec.ts:23]: add type `DeepMocked<ApplicationService>;`

File: alcs-api/src/application/application-decision-meeting/application-decision-meeting.entity.ts [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+    }
+  }
+
+  @AutoMap()
+  @CreateDateColumn({ type: 'timestamptz' })
+  date: Date;
+
+  @AutoMap()
+  @Column()
+  application_uuid: string;
```
Comment (dhaselhan) [inline alcs-api/src/application/application-decision-meeting/application-decision-meeting.entity.ts:21]: Were there issues writing this as applicationUuid?

File: alcs-api/src/application/application-decision-meeting/application-decision-meeting.entity.ts [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+
+  @AutoMap()
+  @ManyToOne(() => Application)
```
Comment (dhaselhan) [inline alcs-api/src/application/application-decision-meeting/application-decision-meeting.entity.ts:24]: Its good to specify the reverse relationship as well since it is setup.
`@ManyToOne(() => Application, (app) => app.decisionMeetings)`

File: alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts (hunk 7 above)
Comment (mhuseinov) [reply alcs-frontend/src/app/services/application/application-decision-meeting/application-decision-meeting.service.ts:53]: should not service method that creates entity to return that entity?

File: alcs-frontend/src/app/features/application/review/decision-meeting/decision-meeting.component.ts (hunk 15 above)
Comment (mhuseinov) [reply alcs-frontend/src/app/features/application/review/decision-meeting/decision-meeting.component.ts:45]: exactly the reason, otherwise I already have all the data there

File: alcs-frontend/src/app/features/application/review/decision-meeting-dialog/decision-meeting-dialog.component.html (hunk 19 above)
Comment (mhuseinov) [reply alcs-frontend/src/app/features/application/review/decision-meeting-dialog/decision-meeting-dialog.component.html]: already implemented that :)

File: alcs-frontend/src/app/features/application/review/decision-meeting/decision-meeting.component.html (hunk 16 above)
Comment (mhuseinov) [reply alcs-frontend/src/app/features/application/review/decision-meeting/decision-meeting.component.html:1]: Not all styling needs to be with margins and paddings. I prefer to add <br> to sections whenever I need to have even spacing between sections. This way the spacing is always the same and no need of additional wrappers or css tweaks

File: alcs-api/src/application/application-decision-meeting/application-decision-meeting.entity.ts (hunk 25 above)
Comment (mhuseinov) [reply alcs-api/src/application/application-decision-meeting/application-decision-meeting.entity.ts:24]: based on the docs it is not a requirement

File: alcs-frontend/src/app/features/application/review/decision-meeting/decision-meeting.component.ts (hunk 15 above)
Comment (dhaselhan) [reply alcs-frontend/src/app/features/application/review/decision-meeting/decision-meeting.component.ts:45]: sounds good

File: alcs-api/src/application/application-decision-meeting/application-decision-meeting.entity.ts (hunk 25 above)
Comment (dhaselhan) [reply alcs-api/src/application/application-decision-meeting/application-decision-meeting.entity.ts:24]: huh I didnt know that, also looks like it doesnt do anything lol.

**Review by dhaselhan — APPROVED** [review]
LGTM 👍
